### PR TITLE
☑️ Refacto "Select All/Unselect all" on indexes

### LINKED ISSUE #4397: ☑️ Enable "Select All" for delete action
```[tasklist]
### Tasks
- [ ] https://github.com/twentyhq/twenty/issues/4406
- [ ] https://github.com/twentyhq/twenty/issues/4409
- [ ] https://github.com/twentyhq/twenty/issues/4408
- [ ] https://github.com/twentyhq/twenty/issues/4407
- [ ] https://github.com/twentyhq/twenty/issues/4412
- [ ] https://github.com/twentyhq/twenty/issues/5169
```

### LINKED ISSUE #5169: Improve the Backend for Mass Record Deletion
# Current Behavior

pg_graphql constraints restrict deletion to a maximum of 30 records per operation.

# Desired Behavior

Implement a loop on the delete request to delete more records in a single operation.

### PR DESCRIPTION
**Pull request by gitstart-twenty**
- Refacto "Select All/Unselect all" on indexes
- Add sequential mass deletion from front end (limited to 10k records)
- Fixed coverage with new unit tests on new useFetchAllRecordIds hook and other utils

Closes #4397
Closes #5169

https://github.com/twentyhq/twenty/assets/26528466/2658ad2c-827e-4670-b42b-3092e268ff32

### CODE REVIEW

**Comment by gitstart-twenty** [comment]
Hey @charlesBochet 
What do you think of this v1?

**Comment by charlesBochet** [comment]
@gitstart-twenty could you rebase your PR on main, tests should be green :)

**Comment by gitstart-twenty** [comment]
> @gitstart-twenty could you rebase your PR on main, tests should be green :)

Alright

**Comment by FelixMalfait** [comment]
Hey, video looks great (nice improvement!) but the total count in the view switcher doesn't get refreshed apparently, we should fix that if that hasn't been fixed already.
(I didn't check the code)

**Comment by lucasbordeau** [comment]
@gitstart-twenty Seems like a good DRY increment for me, this starts to be a big file and difficult to read but it would need a more in depth refactor.

**Comment by lucasbordeau** [comment]
Though I would name it according to its context which is mainly about pagination. Maybe something like useFindManyRecordsPaginationUtils ?

**Comment by gitstart-twenty** [comment]
> Though I would name it according to its context which is mainly about pagination. Maybe something like useFindManyRecordsPaginationUtils ?

Alright

**Comment by gitstart-twenty** [comment]
> Hey, video looks great (nice improvement!) but the total count in the view switcher doesn't get refreshed apparently, we should fix that if that hasn't been fixed already. (I didn't check the code)

Looking into this

**Comment by gitstart-twenty** [comment]
@FelixMalfait 
We're attempting to refetch the records after the import(to get the updated total count) but we keep running into the error below

```diff
 + refetchQueries: [getOperationName(findManyRecordsQuery) ?? ''],
 ```

The error
```
 [
    InternalServerErrorException: GraphQL errors on person: {"message":"Unknown field \"personCollection\" on type Query"}
        at computePgGraphQLError (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/utils/compute-pg-graphql-error.util.ts:52:10)
        at WorkspaceQueryRunnerService.parseResult (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:612:42)
        at WorkspaceQueryRunnerService.findMany (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:116:17)
        at processTicksAndRejections (node:internal/process/task_queues:95:5)
        at field.resolve (/.../node_modules/@envelop/on-resolve/cjs/index.js:36:42)
        at /.../node_modules/graphql-yoga/node_modules/@graphql-tools/executor/cjs/execution/promiseForObject.js:18:35
        at async Promise.all (index 0) {
      path: [ 'people' ],
      locations: [ [Object] ],
      extensions: [Object: null prototype] {}
    }
  ]
```

**Comment by FelixMalfait** [comment]
Not sure sorry!

**Comment by gitstart-twenty** [comment]
> Not sure sorry!

Alright, debugging!

**Comment by gitstart-twenty** [comment]
> Not sure sorry!

Alright!

**Comment by gitstart-twenty** [comment]
> @FelixMalfait We're attempting to refetch the records after the import(to get the updated total count) but we keep running into the error below
> 
> ```diff
>  + refetchQueries: [getOperationName(findManyRecordsQuery) ?? ''],
> ```
> 
> The error
> 
> ```
>  [
>     InternalServerErrorException: GraphQL errors on person: {"message":"Unknown field \"personCollection\" on type Query"}
>         at computePgGraphQLError (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/utils/compute-pg-graphql-error.util.ts:52:10)
>         at WorkspaceQueryRunnerService.parseResult (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:612:42)
>         at WorkspaceQueryRunnerService.findMany (/.../packages/twenty-server/src/engine/api/graphql/workspace-query-runner/workspace-query-runner.service.ts:116:17)
>         at processTicksAndRejections (node:internal/process/task_queues:95:5)
>         at field.resolve (/.../node_modules/@envelop/on-resolve/cjs/index.js:36:42)
>         at /.../node_modules/graphql-yoga/node_modules/@graphql-tools/executor/cjs/execution/promiseForObject.js:18:35
>         at async Promise.all (index 0) {
>       path: [ 'people' ],
>       locations: [ [Object] ],
>       extensions: [Object: null prototype] {}
>     }
>   ]
> ```

Seems to work now!

**Comment by gitstart-twenty** [comment]
Something(new changes) broke the deletion

**Comment by lucasbordeau** [comment]
@gitstart-twenty Do you need something from our side to move forward ?

**Comment by gitstart-twenty** [comment]
> @gitstart-twenty Do you need something from our side to move forward ?

@lucasbordeau Going to update you on this ASAP!

**Comment by FelixMalfait** [comment]
@gitstart-twenty let's try to close this one this week since it's been open for long! Thanks

**Comment by gitstart-twenty** [comment]
> @gitstart-twenty let's try to close this one this week since it's been open for long! Or let us know if we should close it now. Thanks

Hey @FelixMalfait This is going to be completed!

File: packages/twenty-front/src/modules/object-record/hooks/useFindManyRecords.ts [hunk 1]
```diff
@@ -126,111 +98,24 @@ export const useFindManyRecords = <T extends ObjectRecord = ObjectRecord>({
       },
```
Comment (gitstart-twenty) [inline packages/twenty-front/src/modules/object-record/hooks/useFindManyRecords.ts:98]: Line 97 above is the cause of the errors getting thrown by the two actions `front-sb-build` and `front-task (typecheck)`, however, notice that we do not make any changes to it to cause this error!!

Screenshot 2024-06-11 at 18 02 09

**Comment by gitstart-twenty** [comment]
https://github.com/twentyhq/twenty/assets/140154534/8ed841c0-03f4-427f-8ff1-951d516052a9

**Review by lucasbordeau — COMMENTED** [review]
Ok but this part of the code needs to be refactored, though it's not the scope of this PR

**Comment by lucasbordeau** [comment]
@gitstart-twenty I just tested but I have unpredictable results, sometimes the export gets triggered before 100%

Would it be difficult for you to remove the useEffect and use a loop/map that awaits for each lazy fetchMore ?

**Comment by gitstart-twenty** [comment]
> @gitstart-twenty I just tested and I have unpredictable results, sometimes the export gets triggered before 100%
> 
> Would it be difficult for you to remove the useEffect and use a loop/map that awaits for each lazy fetchMore ?

Hey @lucasbordeau 
Let's look into this

**Comment by gitstart-twenty** [comment]
> Would it be difficult for you to remove the useEffect and use a loop/map that awaits for each lazy fetchMore ?
>
Hey @lucasbordeau,

Using a loop has proven tricky, this is because `fetchMoreRecords` completing doesn't necessarily mean the records have also been updated with the new ones. So, much as we can loop and wait for each to complete before running the next one, we will most likely still have incomplete results (tried and tested)!

Consequently, we'd have to watch the records which brings us right back to the `useEffect`

**Comment by gitstart-twenty** [comment]
Perhaps we can look into ensuring the `useEffect` always ensures all records have been updated before running the callback?

**Comment by gitstart-twenty** [comment]
Hey @lucasbordeau 
We resorted to fixing the `useEffect`. The change below seems to make a differenc
Screenshot 2024-06-21 at 14 05 50
e, please test out on your side

**Comment by lucasbordeau** [comment]
@gitstart-twenty That works for me.

Just a minor issue left : the header checkbox should be checked if all rows are selected, not in indeterminate state like now.

**Comment by lucasbordeau** [comment]
And after export, the header checkbox stays at selected state, even if all the rows are unselected.

**Comment by gitstart-twenty** [comment]
Hey @lucasbordeau 
> And after export, the header checkbox stays at selected state, even if all the rows are unselected.

https://github.com/twentyhq/twenty/assets/140154534/e7ae3a44-8c8f-45f1-92e5-2b4026efcc83

**Comment by lucasbordeau** [comment]
When we deselect after select all we should deselect from total, not from loaded rows in front.


https://github.com/twentyhq/twenty/assets/26528466/dbeb9ee2-c2b6-4301-9aaf-e9f6f94a9e79

**Comment by gitstart-twenty** [comment]
> When we deselect after select all we should deselect from total, not from loaded rows in front.
> 
> Please make sure that you get a table with more than 300/400 rows so that you're not having loaded rows = db rows, which hides some edge cases

Interesting! Working on it!

**Comment by gitstart-twenty** [comment]
Hey @lucasbordeau 

We need to support a `not-in` filter in order for this to work efficiently, however, I'm still facing some challenges implementing it!

**Comment by gitstart-twenty** [comment]
Hey @lucasbordeau,

I've noticed that the `pg_graphql` extension being used on the project is not accepting the `nin` filter. However, the [official docs](https://github.com/supabase/pg_graphql/pull/344/files) say it should work! Is this the intended behavior?

I'm trying to use the `nin` filter for search when a user selects multiple items and then deselects some. The `nin` filter would be perfect for this use case. The other option would be to fetch all the selected items and then filter out the deselected ones in the application layer, but this is causing `re-renders` in the `useEffect` making it hard to implement.

**Comment by gitstart-twenty** [comment]
The filter is actually not listed [here](https://supabase.github.io/pg_graphql/api/)

**Comment by gitstart-twenty** [comment]
We can use the `not` operator, my bad!

**Comment by gitstart-twenty** [comment]
The count should be fine now

https://github.com/twentyhq/twenty/assets/140154534/3907eec2-461f-4c39-99d9-95632f218c15

### NOTE

There's a bug when selecting a lot of records, not all might be deleted, the count is also not updating in real-time for many records selected at once 😓

**Comment by gitstart-twenty** [comment]
I've just noticed a new change that limits mutations to `100` records max, this should help us resolve our issue here, `100` records is manageable

**Comment by gitstart-twenty** [comment]
Hey @lucasbordeau I somehow missed this comment, let me update this ASAP!

**Comment by lucasbordeau** [comment]
@gitstart-twenty I'll take on from here as this PR is connected to multiple other refactors and another issue.

**Review by lucasbordeau — APPROVED** [review]

**Pull request merged**